### ISSUE TITLE
ZSTD S/W should be able to decompress a compressed block of size 128KB.

### ISSUE THREAD

**Issue by SmitaKumar** [facebook/zstd#2577]
The RFC has an additional restriction on the Block_Size that is not mentioned in the GITHUB spec. Also, there is an ERRATA in the RFC (https://www.rfc-editor.org/errata_search.php?rfc=8478) for this line 

>  A Compressed_Block has the extra restriction that Block_Size is
>    always strictly less than the decompressed size.  If this condition
>    cannot be respected, the block must be sent uncompressed instead
>    (i.e., treated as a Raw_Block).

ZSTD S/W cannot decompress a compressed block of size 128KB. This restriction isn't apparent from the GITHUB spec.
```
if (srcSize >= ZSTD_BLOCKSIZE_MAX) {
return ERROR(srcSize_wrong);
}
```
GITHUB spec

> Block_Content and Block_Maximum_Size
> 
> The size of Block_Content is limited by Block_Maximum_Size, which is the smallest of:
> 
> Window_Size
> 128 KB
> Block_Maximum_Size is constant for a given frame. This maximum is applicable to both the decompressed size and the compressed size of any block in the frame.
> 
> The reasoning for this limit is that a decoder can read this information at the beginning of a frame and use it to allocate buffers. The guarantees on the size of blocks ensure that the buffers will be large enough for any following block of the valid frame.

**Comment by SmitaKumar** [comment]
I see an errata in the RFC (https://www.rfc-editor.org/errata_search.php?rfc=8478) but also see the following in the ZSTD S/W code,
if (srcSize **>=** ZSTD_BLOCKSIZE_MAX) {
  return ERROR(srcSize_wrong);
 }

**Comment by Cyan4973** [comment]
Hi @SmitaKumar 

This part of the specification has actually been updated.

Due to the way RFC publication works, the updated version of the specification cannot be "republished" as RFC 8478. It has to be a new document, and as a consequence, it has received a new number.
It is [RFC 8878](https://www.rfc-editor.org/rfc/rfc8878.html).

The updated RFC matches the wording of the Github spec (https://www.rfc-editor.org/rfc/rfc8878.html#name-blocks).
A compressed block is allowed to expand the size of input (even though it may seem inefficient), as long as it remains within the limits of a full block, which is consistent with decoder's buffer allocation policy.

Is your question about, specifically, the size 128 KB ?
Indeed, 128 KB fits the definition, so a compressed block of 128 KB should be valid and decodable.
If that's not the case, we need to fix that.
(mind linking to the faulty code line ?)

Presuming the issue is confirmed, 
unfortunately even if it's fixed, it would also mean that a large part of `zstd`'s installed base is only compatible with compressed block sizes of up to `128 KB - 1`. Therefore, even though the spec states that `128 KB` precisely should be supported, I would recommend to stay within the more restricted boundaries, and intentionally not generate compressed blocks of 128 KB, to remain compatible with the largest possible ecosystem.

Note that I can't think of a good reason to generate a compressed block of size `== 128 KB`. Moreover, the reference encoder will never generate a compressed block of that size. So this issue is a corner case, that is unlikely to affect the current ecosystem, but must be cleaned for the sake of a complete and consistent spec.

**Comment by SmitaKumar** [comment]
Is your question about, specifically, the size 128 KB ?
Yes. The ZSTD decompressor reports an error (Decoding error (36) : Src size is incorrect ) for a compressed block of Size 128KB

Agree on your point regarding the installed base. Thank you for the suggestion.

**Comment by terrelln** [comment]
CC @Cyan4973:

This seems to be the offending line:

https://github.com/facebook/zstd/blob/40a718813070e23c3c7a437477a9e984d4de3f8b/lib/decompress/zstd_decompress_block.c#L2018

I don't have time to test & fix before I go, so just adding to our release blocking tasks.

**Comment by Cyan4973** [comment]
Fixed in #3399
